crates/gitbutler-core/src/ops/oplog.rs

```diff
@@ -218,44 +158,50 @@ impl Oplog for Project {
 
         // merge all the branch trees together, this should be our worktree
         // TODO: when we implement sub-hunk splitting, this merge logic will need to incorporate that
-        if head_trees.is_empty() {
+        if head_tree_ids.is_empty() {
             // if there are no applied branches, then it's just the target tree
-            tree_builder.insert("workdir", target_tree_oid, FileMode::Tree.into())?;
-        } else if head_trees.len() == 1 {
+            tree_builder.insert("workdir", target_tree_id, FileMode::Tree.into())?;
+        } else if head_tree_ids.len() == 1 {
             // if there is just one applied branch, then it's just that branch tree
-            tree_builder.insert("workdir", head_trees[0].into(), FileMode::Tree.into())?;
+            tree_builder.insert("workdir", head_tree_ids[0].into(), FileMode::Tree.into())?;
         } else {
             // otherwise merge one branch tree at a time with target_tree_oid as the base
-            let mut workdir_tree_oid = target_tree_oid;
-            let base_tree = repo.find_tree(target_tree_oid)?;
+            let mut workdir_tree_id = target_tree_id;
+            let base_tree = repo.find_tree(target_tree_id)?;
             let mut current_ours = base_tree.clone();
 
-            let head_trees_iter = head_trees.iter();
             // iterate through all head trees
-            for head_tree in head_trees_iter {
-                let current_theirs = repo.find_tree(git2::Oid::from(*head_tree))?;
+            // TODO: This needs a test. Right now only the last merged index is actually used.
```

Comment:
Here we are reconstructing the full working directory state from all of the applied virtual branches merged together. Per definition applied virtual branches are assumed to not conflict with each other (because they are applied, i.e. in the working directory).

Is your comment about the condition in which the merge is not clean or did i not understand it correctly? Maybe there is still something we can do here to code it more defensively

Reply:
<img width="681" alt="Screenshot 2024-05-27 at 19 00 20" src="https://github.com/gitbutlerapp/gitbutler/assets/63622/eced2422-6f49-4537-8972-3153841123c8">

The `workdir_tree_id` is completely decoupled from the input of the merge, which means that only the last pair of trees is actually written.

Now I will try to prove it by having a test that validates the produced tree.

----

Something else I was thinking is that the worktree can be reproduced from state which is stored in the snapshot - only the `is_applied()` information would also be needed.

Then one could save the time to merge everything together and instead do that work only when a restore is requested. And there, maybe existing vbranch functionality can be used, after all a restore would restore the vbranches, and from there it should be possible to trigger an 'apply' of sorts. But… this is nothing more than a flash of an idea, nothing for now or even near-term.

Reply:
I'm a little confused, unless I'm misreading this. Isn't `current_ours` being used the same as `workspace_tree` in the second example?

Reply:
`current_ours` is overwritten with `current_theirs` each iteration, giving it the effect of a sliding window, i.e. `merge(base, a, b)`, then `merge(base, b, c`. The problem seems to be that `merge()` returns an index which, despite being turned into a tree, is entirely ignored on all iterations except the last one.

In the second example, the procedure is different, as it's `merge(base, base, a) -> merge_result1`, then `merge(base, merge_result1, b) -> merge_result_2` and so forth, thus previous merge is actually used.

Maybe you are able to validate oplog workspace restore with multiple applied vbranches real quick? Maybe it works and I am misreading this entirely.

But in any case, I think there should be a test for that.

Reply:
https://github.com/gitbutlerapp/gitbutler/assets/63622/a9db410e-1366-4de9-ad58-f8988463b389


Here is a video showing that only the last two virtual branches will be merged into the workdir state. Now for a test that allows to fix that.

Reply:
okay so definitely a bug! its great that you caught this one before we launched the new release :)